### PR TITLE
WLT-1720: allow selecting tokens not held in wallet in swap input picker

### PR DESCRIPTION
**Pull request by zerts**
## Summary

The SwapForm2 Pay-with selector only listed tokens the wallet holds, so users couldn't pick a token they don't have (e.g. to simulate a swap, or when the holding wallet isn't connected). Now a non-empty search query also runs a backend token search (`useSearchQueryFungibles`, same hook the Receive selector uses, debounced 300ms) and shows matches the wallet does **not** hold as zero-balance rows in a "Tokens not held in wallet" section below the holdings.

Mirrors the web-app implementation (zeriontech/zerion-web-app@f5b4675), which itself was a port of this extension component.

- Each search result expands into one row per tradable implementation chain; a selected network chip narrows results to that chain, "All networks" orders chains chip-order-first, then backend order. Held (chain, fungible) pairs are deduped out of the search section.
- Selecting a not-held token flows through the existing zero-balance synthesis in `useFormPositions`, so the form shows Balance: 0 in red with the shake animation and the quote-level "Insufficient balance" warning instead of blocking the pick.
- `InputPositionSelector.onSelect` signature changed from `(position, selectedTab)` to `(chainId, fungibleId, selectedTab)` since not-held rows have no position (the single caller only used those two fields).
- Skeleton rows while the first search page loads; search state now resets on dialog close (also fixes a latent bug where a stale query kept filtering holdings after reopen).
- Promoted `TokenListSkeleton` from `OutputPositionSelector` to the shared `src/ui/components/PositionSelector/`; added an optional `enabled` flag to `useSearchQueryFungibles`.

Closes WLT-1720

## Test plan

- Open Swap → Pay-with selector → type a token symbol you don't hold (e.g. on a fresh wallet): a "Tokens not held in wallet" section appears below holdings with zero-balance rows per tradable chain.
- Select a network chip → search rows narrow to that chain only; "All networks" shows one row per implementation chain, chip-order first.
- Pick a not-held token → selector closes, token is set as Pay-with, balance shows 0; entering an amount shows the insufficient-balance state instead of blocking.
- Tokens you already hold don't duplicate into the search section.
- While search loads, skeleton rows show instead of "No tokens found"; reopening the dialog starts with a clean query.
- Receive selector behavior unchanged.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by zerts** [comment]
## ✅ QA smoke gate — iteration 1/5

**Result:** `pass`

### Screenshots

**Unlocked overview (persona default, multi-chain portfolio loaded)**

Unlocked overview (persona default, multi-chain portfolio loaded)

**SwapForm2 default state (ETH → wstETH)**

SwapForm2 default state (ETH → wstETH)

**Pay-with selector open, holdings list**

Pay-with selector open, holdings list

**Search "pepe" — "Tokens not held in wallet" section with per-chain rows below held result**

Search "pepe" — "Tokens not held in wallet" section with per-chain rows below held result

**Not-held PEPE (Base) selected as Pay-with, Balance 0**

Not-held PEPE (Base) selected as Pay-with, Balance 0

**Amount entered — red Balance 0 + explicit "Insufficient balance" warning, live quote**

Amount entered — red Balance 0 + explicit "Insufficient balance" warning, live quote

## Summary

Persona "default" (active wallet 0x52Fb…78e59), extension built from PR head 8d5d8916 and
driven headless via seeded persistent context. Popup unlocked, portfolio rendered. Drove
the affected flow: `#/swap-form` → Pay-with selector (Shift+ArrowUp) → typed "pepe" →
the new "Tokens not held in wallet" section rendered below the held result, with verified
Pepe rows expanded one-per-chain (chip-order first: Base, Ethereum, …) → clicked the
first not-held row → selector closed, PEPE (Base) set as Pay-with with Balance: 0 →
entered amount 1 → red "Balance: 0", live quote (0.00000000176 ETH), and the explicit
"Insufficient balance — Input asset balance is not enough to execute a swap" banner with
"Top Up Your Wallet" CTA. This is exactly the acceptance behavior: selection allowed,
explicit error instead of blocking.

## Verification log
- Profile: zerion-wallet-extension — persona "default", no flag/dev-menu override needed
  (`#/swap-form` renders SwapForm2 ungated)
- Built: `npm run build:chrome` → dist/manifest.json OK
- Flows: unlock → `#/swap-form` → Pay-with selector → search "pepe" → select not-held
  PEPE (Base) → enter amount 1 → insufficient-balance state
- Backend: zpi.zerion.io healthy — search results, asset-list fungible resolution, and
  swap quote all returned data (quote visible in 06)
- Console/network: two console `400` resource errors during quote-pair transitions and one
  `stream-swap-quotes … net::ERR_ABORTED` (superseded quote stream on input change) — the
  final state shows a live quote and correct warning, so not flow-breaking; otherwise clean

_Posted by [ai-runner](http://127.0.0.1:4777)._

**Comment by zerts** [comment]
## ✅ Loop review — iteration 1/5

**Verdict:** `approved`

| Severity | Area | Finding |
| --- | --- | --- |
| minor | behavior | Held token searched by contract address appears only in the not-held section (`src/ui/pages/SwapForm2/PositionSelector/InputPositionSelector.tsx:265`) |

## Summary

Reviewed PR #960 at head `8d5d8916197a83381726b491d71874ed3ae3c9f0` against WLT-1720 and the
PRD comment on the issue. The change adds a backend token search to the SwapForm2 Pay-with
selector so tokens not held in the wallet can be selected, surfacing them as zero-balance rows
under a "Tokens not held in wallet" header, with the existing insufficient-balance error state
communicating the problem instead of blocking the pick. Everything in the spec is implemented
and verified working live in the built extension: search section rendering with per-chain row
expansion (chip-order first), network-chip narrowing, held-pair dedup, not-held selection
flowing into Balance: 0 → red balance + "Insufficient balance" banner + live quote, skeleton
loading path, and search reset on dialog close (including the latent stale-query-on-reopen bug
fix). Type-check passes; the single unit-test failure is pre-existing and unrelated. No human
review threads existed on the PR to reconcile. One minor pre-existing edge case noted below;
it does not warrant a fix iteration.

## Findings

### [minor/behavior] Held token searched by contract address appears only in the not-held section
`filteredPositions` filters holdings by `normalizedContains` on fungible name/symbol only
(InputPositionSelector.tsx:265-272), while the backend search also matches contract addresses.
Searching the address of a token the wallet holds therefore hides the held row (name/symbol
don't contain the address) while the search section shows the same token for chains where it
is not held — the held chain's row is invisible and the section implies the user doesn't hold
the token at all. Pre-existing limitation of the holdings text filter (before this PR an
address query simply showed "No tokens found"); the new section only makes it slightly more
visible. Expected: a held token matched by address search would ideally surface its held row
too. Actual: only not-held chain rows appear. Non-blocking edge case.

## Verification log

- Code review: full diff (5 files, +274/−78) read against the Linear spec, PRD comment, and
  web-app reference behavior. `VirtualizedTokenList` natively supports `header` items +
  `renderHeader`; the promoted shared `TokenListSkeleton`'s CSS classes all exist in
  `src/ui/components/PositionSelector/styles.module.css`; `useSearchQueryFungibles`'s new
  second param is optional so the old-form caller (`MarketAssetSelect`) is unaffected;
  `InputPositionSelector`'s only caller (`InputPosition.tsx`) was updated for the new
  `onSelect` signature; `useFormPositions.useEmptyPosition` already synthesizes the
  zero-balance input position the not-held pick relies on. Dedup keys on
  `(chainId, fungibleId)`, matching position identity. No animation code in diff.
- Human PR feedback: no review threads (resolved or open) and no human conversation comments
  existed on the PR — only the Linear linkback, a build artifact, and the ai-runner QA gate.
  Nothing to reconcile.
- Tests: `tsc --noEmit` clean. `npm run test:unit`: 363/364 passed; the 1 failure
  (`src/modules/ethereum/transactions/gasPrices/estimateNetworkFee.test.ts`, small numeric
  drift in Optimism fee estimation) is in a module untouched by this diff, last modified in
  commit `17e93271` well before this branch — pre-existing, non-blocking. Repo Playwright
  e2e suite (`test:e2e`) not run — user instructed mid-review to skip e2e for this run.
- Live: built `dist/` via `npm run build:chrome` from the PR head and drove the extension
  popup headless (Playwright chromium, persistent context seeded with the "default" persona,
  viewport 440×900). Flows: unlock → `#/swap-form` → Pay-with selector (Shift+ArrowUp) →
  typed "pepe" → "Tokens not held in wallet" section rendered below the held result with
  verified Pepe rows expanded per chain, chip-order first (Base, Ethereum, then backend
  order), no fiat/quantity on not-held rows → Base chip narrowed the section to a single
  Pepe-on-Base row → back to All → clicked the first not-held row → selector closed, PEPE
  (Base) set as Pay-with, Balance: 0 → entered amount 1 → red "Balance: 0", live quote
  (0.00000000176 ETH), orange "Insufficient balance — Input asset balance is not enough to
  execute a swap" banner with "Top Up Your Wallet" CTA → reopened selector: search input
  empty, full holdings restored (reset-on-close verified) → searched "eth" and scrolled the
  full virtualized list (142 rows): held section 71 rows / not-held section 71 rows with no
  observable duplication (ETH held on ~17 chains, only ~10 not-held ETH rows — consistent
  with held-pair dedup; the raw "DUPLICATES" output from my scroll script is an extraction
  artifact: chain `alt` text came back empty so rows matched on name alone) → Receive
  selector (Shift+ArrowDown) renders unchanged (Popular/Others sections). Skeleton state not
  visually captured (search resolved faster than the debounce window); code path reviewed
  instead. Screenshots: /tmp/loop-review-wlt1720/shots (01–12).
- Console/network: two bare `Failed to load resource: 400` console errors during quote-pair
  transitions and one `stream-swap-quotes … net::ERR_ABORTED` (superseded quote stream) —
  both documented pre-existing noise for this form per the QA profile; final rendered state
  showed a live quote. Otherwise clean; zpi.zerion.io healthy throughout.

_Posted by [ai-runner](http://127.0.0.1:4777)._

File: src/ui/pages/SwapForm2/PositionSelector/InputPositionSelector.tsx [hunk 1]
```diff
@@ -180,16 +273,91 @@ export function InputPositionSelector({
     return result.sort((a, b) => (b.amount.value || 0) - (a.amount.value || 0));
   }, [tradablePositions, selectedNetwork, searchValue]);
 
-  const virtualItems = useMemo<VirtualListItem<FungiblePosition>[]>(
+  const { fungibles: searchFungibles, queryData: searchQueryData } =
```
Comment (zerts) [inline src/ui/pages/SwapForm2/PositionSelector/InputPositionSelector.tsx:296]: the minor problem from the qa report is very important, lets fix it!

Comment (zerts) [reply src/ui/pages/SwapForm2/PositionSelector/InputPositionSelector.tsx:296]: Fixed in c3de8858 — held positions now also match the search query against the fungible id and all implementation addresses (min 4 chars, so single hex chars don't defeat the name/symbol filter). Pasting a held token's contract address now surfaces its held row above the "Tokens not held in wallet" section.

**Comment by zerts** [comment]
## ✅ QA smoke gate — iteration 1/5

**Result:** `pass`

### Screenshots

**Unlocked overview (persona default, multi-chain portfolio loaded)**

Unlocked overview (persona default, multi-chain portfolio loaded)

**SwapForm2 default state (ETH → wstETH)**

SwapForm2 default state (ETH → wstETH)

**Pay-with selector open, holdings list**

Pay-with selector open, holdings list

**Search "pepe" — "Tokens not held in wallet" section with per-chain rows below held result**

Search "pepe" — "Tokens not held in wallet" section with per-chain rows below held result

**Not-held PEPE selected as Pay-with, Balance 0**

Not-held PEPE selected as Pay-with, Balance 0

**Amount entered — red Balance 0 + explicit "Insufficient balance" warning, live quote**

Amount entered — red Balance 0 + explicit "Insufficient balance" warning, live quote

**Review fix — held USDC searched by contract address, held rows (with balances) above the not-held section**

Review fix — held USDC searched by contract address, held rows (with balances) above the not-held section

**Held USDC picked from address search — real Balance 6.532, live quote, Swap Now**

Held USDC picked from address search — real Balance 6.532, live quote, Swap Now

## Summary

Persona "default" (active wallet 0x52Fb…78e59), extension built from PR head c3de8858
(iteration 2: adds the review fix — held positions now match search queries against their
contract addresses) and driven headless via seeded persistent context. Popup unlocked,
portfolio rendered. Re-drove the original flow: `#/swap-form` → Pay-with selector
(Shift+ArrowUp) → "pepe" → "Tokens not held in wallet" section below the held result →
selected not-held Pepe → Balance: 0 → amount 1 → red balance + "Insufficient balance"
banner with live quote. Then drove the review-fix scenario: reopened the selector and
pasted the contract address of held Arbitrum USDC (0xaf88d065…5831, captured from the
extension's own `wallet/get-simple-positions/v1` response) — held USDC rows rendered at
the top WITH fiat values/quantities ($6.53 / 6.532 USDC first), the "Tokens not held in
wallet" header below them; selecting the held row set Pay-with USDC with real Balance
6.532 and a live 1-USDC → 0.00052 ETH quote. Before this fix an address query hid all
held rows.

## Verification log
- Profile: zerion-wallet-extension — persona "default", no flag/dev-menu override needed
  (`#/swap-form` renders SwapForm2 ungated)
- Built: `npm run build:chrome` from c3de8858 → dist/manifest.json OK
- Flows: unlock → `#/swap-form` → Pay-with selector → search "pepe" → select not-held
  Pepe → amount 1 → insufficient-balance state → reopen selector → paste held USDC
  (arbitrum) contract address → held rows above header → select → real balance + quote
- Numeric assertion (address search): `{"hasHeader":true,"totalRows":14,
  "heldRowsAboveHeader":14,"firstHeldRowText":"USDC MCAP $73.0B $6.53 6.532 USDC"}` —
  every rendered USDC row above the not-held header carries fiat value + quantity
  (held rows), header present below for not-held chains
- Backend: zpi.zerion.io healthy — positions, search, asset-list resolution, and swap
  quotes all returned data (quotes visible in 06 and 08)
- Console/network: three bare `400` console errors and two
  `stream-swap-quotes … net::ERR_ABORTED` during quote-pair transitions (documented
  profile noise; one extra pair transition this run from the added scenario) — final
  states show live quotes, not flow-breaking; otherwise clean

_Posted by [ai-runner](http://127.0.0.1:4777)._

**Comment by zerts** [comment]
## ✅ Loop review — iteration 1/5

**Verdict:** `approved`

## Summary

Reviewed PR #960 at head `c3de88584910c6991b847749a6625acf06ab8f28`. This round centered on
the human review threads: there is exactly one (zerts, on InputPositionSelector.tsx:296,
resolved), which escalated the previous review's minor finding — a held token searched by its
contract address appeared only under "Tokens not held in wallet" while its held row stayed
hidden — and asked for a fix. The fix genuinely landed at the current head as commit
`c3de8858`: a `matchesAddressQuery` helper is OR-ed into the holdings filter, matching the
lowercased query against the fungible id and every implementation address (guarded to queries
of ≥4 chars so short hex fragments don't defeat the name/symbol filter, and handling null
addresses on native assets). Verified live in the built extension: pasting USDT's Ethereum
contract address now shows 11 held Tether USD rows with balances above the "Tokens not held
in wallet" section, with not-held USDT chains below and no duplication; symbol search ("pepe")
is unregressed. Type-check clean, unit tests show only the known pre-existing failure. No
remaining findings.

## Findings

None.

## Verification log

- Human threads: 1 thread total (resolved, not outdated) — change requested by zerts was
  verified present and working at head; the implementer's reply accurately describes the
  code. No other human comments amend the spec.
- Code review: delta since previously-approved `8d5d8916` is the single commit `c3de8858`
  (+21/−1 in InputPositionSelector.tsx). Call site lowercases the query before matching;
  `AssetImplementation.address` nullability is guarded; matching is additive (OR) so
  name/symbol behavior is unchanged; dedup of held pairs out of the search section is
  untouched and still keyed on (chainId, fungibleId). No animation code in diff.
- Tests: `tsc --noEmit` clean. `npm run test:unit` run three times: two runs show only the
  known pre-existing failure (`gasPrices/estimateNetworkFee.test.ts`, module untouched by
  this PR, 363/364 passing); one run additionally flaked on a second suite that did not
  reproduce in either subsequent run. Repo Playwright e2e suite not run: it covers only
  onboarding/login/settings/ledger snapshot screens (nothing near the swap form), and the
  user previously instructed skipping it in this loop.
- Live: rebuilt `dist/` via `npm run build:chrome` at head and drove the popup headless
  (Playwright chromium, persistent context seeded with the "default" persona, 440×900).
  Flows: unlock → `#/swap-form` → Pay-with selector → pasted
  `0xdac17f958d2ee523a2206206994597c13d831ec7` (USDT Ethereum) → 11 held Tether USD rows
  with fiat/quantity rendered above the "Tokens not held in wallet" header (row-dump
  confirmed header at list index 11 with zero-balance USDT rows after it) → cleared query,
  typed "pepe" → held Young Peezy row + per-chain not-held Pepe rows, identical to the
  previously approved behavior. Screenshots: /tmp/loop-review-wlt1720/shots-r2 (01–02),
  both read and judged.
- Console/network: one bare `Failed to load resource: 400` console line during form load
  (documented pre-existing noise for this form); no failed requests. zpi.zerion.io healthy.

_Posted by [ai-runner](http://127.0.0.1:4777)._
